File: src/main/java/tools/important/tankslua/gui/EvalBox.java
package tools.important.tankslua.gui;

import party.iroiro.luajava.Lua;
import party.iroiro.luajava.lua54.Lua54;
import tanks.Game;
import tanks.gui.Button;
import tanks.gui.TextBox;
import tanks.gui.screen.Screen;
import tools.important.tankslua.Notification;
import tools.important.tankslua.SafeLuaRunner;
import tools.important.tankslua.TanksLua;

public class EvalBox {
    public Lua luaState;
    public TextBox evalCodeBox;
    public Button evalRunButton;

    public void draw() {
        evalCodeBox.draw();
        evalRunButton.draw();
    }
    public void update() {
        evalCodeBox.update();
        evalRunButton.update();
    }
    public EvalBox() {
        Screen screen = Game.screen;
        luaState = new Lua54();
        luaState.openLibraries();
        TanksLua.initializeState(luaState);

        evalCodeBox = new TextBox(screen.centerX, screen.objYSpace, screen.objWidth * 4, screen.objHeight, "Lua Code", () -> {
        }, "");
        evalCodeBox.allowLetters = true;
        evalCodeBox.allowSpaces = true;
        evalCodeBox.enableCaps = true;
        evalCodeBox.enablePunctuation = true;

        evalCodeBox.maxChars = 1000; // who's gonna write lua code longer than 1000 characters

        SafeLuaRunner runner = TanksLua.tanksLua.runner;

        evalRunButton = new Button(screen.centerX - screen.objXSpace * 1.37, screen.objYSpace * 2, screen.objWidth, screen.objHeight, "Evaluate", () -> {
            String code = evalCodeBox.inputText;
            SafeLuaRunner.LuaResult loadResult = runner.safeLoadString(luaState, code, "evalbox");
            final double youFuckedUpSecondsPerCharacter = 0.1;

            if (loadResult.status != Lua.LuaError.OK) {
                String error = loadResult.errorMessage;
                new Notification(Notification.NotificationType.WARN, youFuckedUpSecondsPerCharacter * error.length(), "Your code failed to load! " + error);
                return;
            }

            SafeLuaRunner.LuaResult callResult = runner.safeCall(loadResult.returns[0]);

            if (callResult.status != Lua.LuaError.OK) {
                String error = callResult.errorMessage;
                new Notification(Notification.NotificationType.WARN, youFuckedUpSecondsPerCharacter * error.length(), "Your code failed to run! " + error);
                return;
            }

            new Notification(Notification.NotificationType.INFO, 1, "Successfully ran your code!");
        });
    }
}

File: src/tools/important/tankslua/gui/EvalBox.kt
package tools.important.tankslua.gui

import party.iroiro.luajava.Lua
import party.iroiro.luajava.LuaException
import party.iroiro.luajava.luajit.LuaJit
import party.iroiro.luajava.value.LuaValue
import tanks.Drawing
import tanks.gui.Button
import tanks.gui.TextBox
import tools.important.tankslua.initialize

class EvalBox {
    val luaState: Lua = LuaJit().apply { initialize() }
    var lastEvalResult: String? = null

    val codeBox: TextBox = TextBox(
        Drawing.drawing.interfaceSizeX * 0.5,
        Drawing.drawing.interfaceSizeY * 0.1,
        Drawing.drawing.objWidth * 3,
        Drawing.drawing.objHeight,
        "Lua Code",
        {},
        ""
    ) .apply {
        allowAll = true
        maxChars = Int.MAX_VALUE // who in their right mind is gonna write lua code longer than this
    }
    val evalButton: Button = Button(
        Drawing.drawing.interfaceSizeX * 0.5,
        Drawing.drawing.interfaceSizeY * 0.1 + Drawing.drawing.objHeight * 1.2,
        Drawing.drawing.objWidth,
        Drawing.drawing.objHeight,
        "Evaluate",
        fun() {
            try {
                val inputText = codeBox.inputText

                val code = if (inputText.startsWith('='))
                    "return "+inputText.substring(1)
                else
                    inputText

                luaState.load(code)
            } catch (e: LuaException) {
                Notification("Your code ran into an issue loading: ${e.message}", NotificationType.ERR)
                return
            }

            val function: LuaValue = luaState.get()

            lastEvalResult = try {
                val results = function.call()

                if (results.size > 0) {
                    results[0].toString()
                } else {
                    "<no return>"
                }
            } catch (e: LuaException) {
                Notification("Your code ran into an issue running: ${e.message}", NotificationType.ERR)
                return
            }

            Notification("Evaluation success!", NotificationType.INFO)
        }
    )

    fun draw() {
        evalButton.draw()
        codeBox.draw()
        val drawing = Drawing.drawing!!

        if (lastEvalResult != null) {
            drawing.setColor(0.0, 0.0, 0.0)
            drawing.setInterfaceFontSize(20.0)
            drawing.drawInterfaceText(codeBox.posX - codeBox.sizeX/2, codeBox.posY + codeBox.sizeY, lastEvalResult)
        }
    }

    fun update() {
        codeBox.update()
        evalButton.update()
    }
}